### DISCUSSION TITLE
Audit events

### DISCUSSION THREAD

**Discussion by yaleman** [kanidm/kanidm#1455] [Ideas]
There's a need for an internal method of handling audit events for various uses:

- #634 Notifying a user of a new MFA device
- there was a discussion about a pub/sub event bus for tools to pull from 
- ... other things!

My initial take is a few things need to be worked out:

- a lightweight schema for events needs to be found (because reinventing the wheel is the worst)
- storing it internally in the k-v store as a timestamped event 
- internal hooks for creating events based on the schema
- internal (and REST) methods for pulling events based on filters (typically "everything matching x filter since y time")
- a way of cleaning them up after a given retention period

**Comment by yaleman** [comment]
As far as schema's concerned, OCSF is a new one that a load of security companies are working together on - https://github.com/ocsf/ocsf-schema

**Reply by Firstyear** [reply]
Yeah, I saw a talk about this at EO the other day! I was hoping to make this structured and object based for events so looking at ocsf is probably a great idea

**Reply by yaleman** [reply]
I'm going to poke around at codegen'ing the schema into usable objects, more for my own learning but it'll be usable here too.

**Comment by Firstyear** [comment]
A question is if we want these events to end up globally stored, or if we want them to end up "per server". I think per server makes more sense for this? Like a remote way to pull logs?

**Reply by yaleman** [reply]
I'd say per-server makes more sense at least initially - one of the modules *could* be some kind of "push to x server" but... eh?

**Reply by Firstyear** [reply]
I think actually per server is a hard requirement given that if we have read only nodes they won't be able to ship events back so having events "pulled" from the consuming source is a better way to go.

**Comment by dvv** [comment]
Hi!

I wonder if it could be an [eventsource](https://html.spec.whatwg.org/multipage/server-sent-events.html#eventsource) of patches in [JSON Patch](https://www.rfc-editor.org/rfc/rfc6902.html) format? I believe it combines well with KV nature of storage.

TIA
